### PR TITLE
WeChat

### PR DESCRIPTION
**Pull request by dustinblackman**
## Overview

Adds Wechat. Depends on https://github.com/broidHQ/integrations/pull/75

### CODE REVIEW

File: broid-wechat/README.md [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+```
+
+**Options available**
+
+| name              | Type     | default    | Description  |
+| ----------------- |:--------:| :--------: | --------------------------|
+| serviceID         | string   | random     | Arbitrary identifier of the running instance |
+| logLevel          | string   | `info`     | Can be : `fatal`, `error`, `warn`, `info`, `debug`, `trace` |
+
+
```
Comment (killix) [inline broid-wechat/README.md]: 👮

File: broid-wechat/README.md [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+  }
+};
+
+wechat.send(formatted_message)
+  .then(() => console.log("ok"))
+  .catch(err => console.error(err));
+```
+
+
+
```
Comment (killix) [inline broid-wechat/README.md]: 👮

File: broid-wechat/README.md [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+{
+  "@context": "https://www.w3.org/ns/activitystreams",
+  "type": "Create",
+  "generator": {
+    "id": "f6e92eb6-f69e-4eae-8158-06613461cf3a",
+    "type": "Service",
+    "name": "wechat"
+  },
+  "object": {
+    "type": "Video",
```
Comment (killix) [inline broid-wechat/README.md]: Audio clip doesn't use the `Audio` type?

File: broid-wechat/lib/test/fixtures/broid/audio.json [hunk 4]
```diff
@@ -0,0 +1,3 @@
+{"@context":"https://www.w3.org/ns/activitystreams","actor":{"id":"test_id","name":"My Name","type":"Person"},"generator":{"id":"test_wechat_service","name":"wechat","type":"Service"},"object":{"id":"6394947961882267211","type":"Audio","url":"http://file.api.wechat.com/cgi-bin/media/get?access_token=test&media_id=64l95jXbkox-iLHo2_2wjNTu5Vs7JBxGkH1b57sqCsvC-zduQEYxC3WLo-IUI2CC"},"published":1488939850,"target":{"id":"test_account_id","name":"test_account_id","type":"Person"},"type":"Create"} 
+
```
Comment (killix) [inline broid-wechat/lib/test/fixtures/broid/audio.json]: 👮

File: broid-wechat/lib/test/fixtures/broid/image.json [hunk 5]
```diff
@@ -0,0 +1,4 @@
+{"@context":"https://www.w3.org/ns/activitystreams","actor":{"id":"test_id","name":"My Name","type":"Person"},"generator":{"id":"test_wechat_service","name":"wechat","type":"Service"},"object":{"id":"6394947425011355184","type":"Image","url":"http://mmbiz.qpic.cn/mmbiz_jpg/AqnhFKzDorO8QTbHzrAehzXy1TMyCNEWThxvsQ9fhibEXEMhmkc2ZyLqWNXiavQpgOn8iaN0uxw0roawSjnojibGrQ/0"},"published":1488939725,"target":{"id":"test_account_id","name":"test_account_id","type":"Person"},"type":"Create"}
+
+
```
Comment (killix) [inline broid-wechat/lib/test/fixtures/broid/image.json]: 👮

Comment (killix) [inline broid-wechat/lib/test/fixtures/broid/image.json]: basically same for all fixtures.

File: broid-wechat/package.json [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    "express": "^4.14.1",
+    "express-xml-bodyparser": "^0.3.0",
+    "fs-extra": "^2.0.0",
+    "mimetype": "^0.0.8",
+    "node-uuid": "^1.4.7",
+    "ramda": "^0.23.0",
+    "request": "^2.80.0",
+    "rxjs": "^5.0.2",
+    "tmp": "^0.0.31",
+    "wechat-api": "git+ssh://git@github.com:node-webot/wechat-api.git#295ad45e199ef93610c4d3a9350de6992a932d8c"
```
Comment (killix) [inline broid-wechat/package.json]: woua weird..

File: broid-wechat/README.md [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+wechat.connect()
+  .subscribe({
+    next: data => console.log(data),
+    error: err => console.error(`Something went wrong: ${err.message}`),
+    complete: () => console.log('complete'),
+  });
+```
+
+**Options available**
+
```
Comment (killix) [inline broid-wechat/README.md]: missing the Http (`IAdapterHTTPOptions`) options

File: broid-wechat/src/core/adapter.ts [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+
+    if (!this.appID) {
+      throw new Error("appID must be set");
+    }
+    if (!this.appSecret) {
+      throw new Error("appSecret must be set");
+    }
+
+    this.client = Promise.promisifyAll(new WeChat(this.appID, this.appSecret));
+    this.client.setEndpoint("api.wechat.com");
```
Comment (killix) [inline broid-wechat/src/core/adapter.ts]: may be we can put the endpoint editable via the options and keep `api.wechat.com` as default?

File: broid-wechat/src/core/adapter.ts [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+  public send(data: ISendParameters): Promise<Object | Error> {
+    this.logger.debug("sending", { message: data });
+
+    return broidSchemas(data, "send")
+      .then(() => {
+        switch (data.object.type) {
+          case "Note":
+            return this.client.sendTextAsync(data.to.id, data.object.content);
+          case "Audio":
+            return this.uploadFile(data.object.url, "voice", "audio.amr")
```
Comment (killix) [inline broid-wechat/src/core/adapter.ts]: The JSON take optional name field. We can use it at the name of the file with the extension by testing the mime type of the url.

File: broid-wechat/src/core/parser.ts [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+  public serviceID: string;
+  private logger: Logger;
+  private userCache: Object;
+  private wechatClient: any;
+
+  constructor(wechatClient: any, serviceID: string, logLevel: string) {
+    this.generatorName = "wechat";
+    this.serviceID = serviceID;
+    this.logger = new Logger("parser", logLevel);
+    this.userCache = {};
```
Comment (killix) [inline broid-wechat/src/core/parser.ts]: i used `new Map();`

File: broid-wechat/src/core/parser.ts [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+        };
+
+        return message;
+      });
+  }
+
+  private parseImage(normalized: any): Promise<IActivityStream> {
+    return this.createActivityStream(normalized)
+      .then((message: IActivityStream) => {
+        message.object = {
```
Comment (killix) [inline broid-wechat/src/core/parser.ts]: check here https://github.com/broidHQ/integrations/blob/8f6aec36c5fd3eaa2917e459491e555da2a657f1/broid-discord/src/core/parser.ts#L104

File: broid-wechat/src/test/parser.ts [hunk 12]
```diff
@@ -0,0 +1,46 @@
+import test from "ava";
+import * as bluebird from "bluebird";
+import * as glob from "glob";
+import * as path from "path";
+
+import Parser from "../core/parser";
+
+const RESPONSE_FIXTURES: any = {};
+glob.sync(path.join(__dirname, "./fixtures/wechat/*.json")).forEach((file: string) => {
```
Comment (killix) [inline broid-wechat/src/test/parser.ts]: nice!

**Comment by killix** [comment]
PR need to be on devel. Like that we can bump the package version directly on devel and make a PR to master and publish the package.

File: broid-wechat/package.json (hunk 6 above)
Comment (dustinblackman) [inline broid-wechat/package.json]: Their NPM package doesn't contain what should of been released based on Github, they pushed outdated files.

File: broid-wechat/src/core/adapter.ts (hunk 8 above)
Comment (dustinblackman) [inline broid-wechat/src/core/adapter.ts]: Actually, this is not needed in the end.

File: broid-wechat/README.md (hunk 3 above)
Comment (killix) [inline broid-wechat/README.md]: Audio clip doesn't use the Audio type?

**Pull request merged**